lib/Dialect/TTIR/IR/TTIROpsInterfaces.cpp

```diff
@@ -17,37 +17,33 @@
 #include "llvm/ADT/SmallVector.h"
 
 mlir::LogicalResult
-mlir::tt::ttir::detail::verifyElementwiseOp(mlir::Operation *op) {
+mlir::tt::ttir::detail::verifyBroadcastable(mlir::Operation *op) {
+  const auto getShape = [](const Value val) {
+    return mlir::cast<mlir::RankedTensorType>(val.getType()).getShape();
+  };
+
+  const auto operandSegmentSizes =
```

Comment:
I can overuse `auto` sometimes and I am aware that it can hinder readability sometimes (language servers tends to mitigate this biggest disadvantage nowadays). The biggest advantage (besides the laziness, can't lie about that) is the form of polymorphism that auto gives us. In this example I don't care if the type of operandSegmentSizes is mlir::DenseI32Array, llvm::ArrayRef, llvm::SmallVector, std::vector, std::array, as long as it has indexing operation and size() member function (with reasonable semantics) it's good to go, and I don't have to change anything if the underlying type changes, because I'm committed only to the interface, not the type.

Reply:
I see your point about not worrying about the underlying type, but do you realistically expect the API to change? I’d also argue that this type of polymorphism can lead to bugs in certain scenarios. For example, imagine you initially use auto with a SmallVector and rely on its indexing operator and size() functionality. Later, the API changes and returns a DenseMap instead, where the indexing operator behaves differently from SmallVector.

I understand the value of auto for lambdas, iterators, and constructor expressions, but beyond that, it feels a bit excessive to me.

Reply:
I don't disagree that it doesn't have disadvantages (basically it's the static version of the duck typing), I just think that benefits that auto brings outweights those disadvantages. The main point is that interfaces are usually more stable than types, your producer is much more likely to change type than interface. Now, in these concrete example I don't expect it to change since the RHS already mentions concrete type in the template, but you can argue that's already a reason in itself to not mention it again in the declaration.
I'm not religious about using `auto`, I just think it was a nice addition to C++ for a lot of reasons, we should just make a project guidelines when it's ok to use it and when it's not.

Reply:
I wasn’t focusing on this specific type, but the fact that the entire function was using auto is what caught my attention...